pkg/controllermanager/controller/managedseedset/managedseedset_actuator.go

```diff
@@ -0,0 +1,414 @@
+	status.Replicas = int32(len(replicas))
+	status.ReadyReplicas = int32(len(readyReplicas))
+
+	// Determine the actual and target replica counts
+	count := len(replicas)
+	targetCount := 0
+	if set.DeletionTimestamp == nil {
+		targetCount = int(*set.Spec.Replicas)
+	}
+
+	// Determine whether scaling out or in
+	scalingOut, scalingIn := count < targetCount, count > targetCount
+
+	// Reconcile the pending replica, if any
+	if pendingReplica != nil {
+		if pending, err := a.reconcileReplica(ctx, set, status, pendingReplica, scalingIn); err != nil || pending {
+			return status, err
+		}
+	}
+
+	switch {
+	case scalingOut:
+		// Initialize a new replica and create its shoot
+		if err := a.createReplica(ctx, set, status, int(status.NextReplicaNumber)); err != nil {
+			return status, err
+		}
+
+		// Increment Replicas and NextReplicaNumber in status
+		status.Replicas++
+		status.NextReplicaNumber++
+
+		return status, nil
+
+	case scalingIn:
+		// Determine the replica to be deleted
+		// From all deletable replicas, choose the one with lowest priority
+		if len(deletableReplicas) == 0 {
+			return status, fmt.Errorf("no deletable replicas found")
+		}
+		sort.Sort(ascendingPriority(deletableReplicas))
+		r := deletableReplicas[0]
+
+		// Delete the replica's managed seed (if it exists), or its shoot (if not)
+		if err := a.deleteReplica(ctx, set, status, r); err != nil {
+			return status, err
+		}
+
+		// Decrement ReadyReplicas in status
+		if replicaIsReady(r) {
+			status.ReadyReplicas--
+		}
+
+		return status, nil
+	}
+
+	// Reconcile postponed replicas
+	for _, r := range postponedReplicas {
+		if pending, err := a.reconcileReplica(ctx, set, status, r, scalingIn); err != nil || pending {
+			return status, err
+		}
+	}
+
+	a.getLogger(set).Debugf("Nothing to do")
+	status.PendingReplica = nil
+	return status, nil
+}
+
+func (a *actuator) reconcileReplica(
+	ctx context.Context,
+	set *seedmanagementv1alpha1.ManagedSeedSet,
+	status *seedmanagementv1alpha1.ManagedSeedSetStatus,
+	r Replica,
+	scalingIn bool,
+) (bool, error) {
+	replicaStatus := r.GetStatus()
+
+	switch {
+	case replicaStatus == StatusShootReconcileFailed && !scalingIn:
+		// This replica's shoot reconciliation has failed, retry it if max retries is not yet reached
+		retries := getPendingReplicaRetries(status, r.GetName(), seedmanagementv1alpha1.ShootReconcilingReason)
+		if int(retries) < *a.cfg.MaxShootRetries {
+			a.infoEventf(set, "Retrying shoot %s reconciliation", r.GetFullName())
+			if err := r.RetryShoot(ctx, a.gardenClient.Client()); err != nil {
+				return false, err
+			}
+			updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.ShootReconcilingReason, pointer.Int32Ptr(retries+1))
+		} else {
+			a.infoEventf(set, "Not retrying shoot %s reconciliation since max retries has been reached", r.GetFullName())
+			updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.ShootReconcileFailedReason, nil)
```

Comment:
IISIC, this call will clear `status.pendingReplia.retries` (set to `nil`), which in turn makes the controller retry the shoot reconciliation again, as soon as the set is enqueued again. 
Is this correct or am I missing something?
```suggestion
			updatePendingReplica(status, r.GetName(), seedmanagementv1alpha1.ShootReconcileFailedReason, &retries)
```
Same would apply to the failed deletion case below.

Reply:
Yes, you are right, fixed.